posts/2018-11-02-semirings.md

```diff
@@ -0,0 +1,435 @@
+Just like with numeric types the laws of `Semiring` state that multiplication has to distribute over addition and multiplying a value with the additive identity (i.e. zero) absorbs the value and becomes zero.
+
+There are different ways to encode this as type classes and different libraries handle this differently, but let's look at how the `typelevel/algebra` project handles this.
+Specifically, it defines a separate `AdditiveSemigroup` and `MultiplicativeSemigroup` and goes from there.
+
+```tut:silent
+import simulacrum._
+
+@typeclass trait AdditiveSemigroup[A] {
+  def +(x: A)(y: A): A
+}
+
+@typeclass trait AdditiveMonoid[A] extends AdditiveSemigroup[A] {
+  def zero: A
+}
+
+@typeclass trait MultiplicativeSemigroup[A] {
+  def *(x: A)(y: A): A
+}
+
+@typeclass trait MultiplicativeMonoid[A] extends MultiplicativeSemigroup[A] {
+  def one: A
+}
+```
+
+A `Semiring` is then just an `AdditiveMonoid` coupled with a `MultiplicativeMonoid` with the following extra laws:
+
+1. Additive commutativity. i.e. `x + y === y + x`
+2. Right Distributivity. i.e. `(x + y) * z === (x * z) + (y * z)`
+3. Left Distributivity. i.e. `x * (y + z) === (x * y) + (x * z)`
+4. Right absorption. i.e. `x * zero === zero`
+5. Left absorption. i.e. `zero * x === zero`
+
+To define it as a typeclass, we simply extend from both additive and multiplicative monoid:
+
+```tut:silent
+@typeclass trait Semiring[A] extends MultiplicativeMonoid[A] with AdditiveMonoid[A]
+```
+
+Now we have a `Semiring` class, that we can use with the various numeric types like `Int`, `Long`, `BigDecimal` etc, but what else is a `Semiring` and why dedicate a whole blog post to it?
+
+It turns out a lot of interesting things can be `Semiring`s, including `Boolean`s, `Set`s and [animations](https://bkase.github.io/slides/algebra-driven-design/#/).
+One very interesting thing, I'd like to point out is the fact that we can form a Semiring homomorphism from types to their number of possible inhabitants.
+What the hell is that?
+Well, bear with me for a while and I'll try to explain step by step.
+
+Okay, so let's start with what I mean by possible inhabitants.
+Every type has a specific number of values it can possibly have, e.g. a `Boolean` has 2 possible inhabitants, because it has two possible values: `true` and `false`.
+
+So `Boolean` has two, how many do other primitive types have?
+`Byte` has 2^8, `Short` has 2^16, `Int` has 2^32 and `Long` has 2^64.
+So far so good, that makes sense, what about something like `String`?
+`String` is an unbounded type and therefore theoretically has infinite number of different inhabitants (practically of course, we don't have infinite memory, so the actual number may vary depending on your system).
+
+For what other types can we determine their number of possible inhabitants? 
+Well a couple of easy ones are `Unit`, which has exactly one value it can take and also `Nothing`, which has 0 possible values.
```

Comment:
I think `Nothing` warrants some more explanation here. Readers may not be aware of how it works. Also, `null` is generally ignored if you want to play by the rules.